std/process.d

```diff
@@ -1034,54 +1034,108 @@ private Pid spawnProcessPosix(scope const(char[])[] args,
-                auto pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
-                if (pfds is null)
-                {
-                    abortOnError(forkPipeOut, InternalError.malloc, .errno);
-                }
-                foreach (i; 0 .. maxToClose)
+                // We read from /dev/fd or /proc/self/fd only if the limit is high enough
+                if (r.rlim_cur > 128*1024)
                 {
-                    pfds[i].fd = i + 3;
-                    pfds[i].events = 0;
-                    pfds[i].revents = 0;
+                    // Try to open the directory /dev/fd or /proc/self/fd
+                    dir = opendir("/dev/fd");
+                    if (dir is null) dir = opendir("/proc/self/fd");
                 }
-                if (poll(pfds, maxToClose, 0) >= 0)
+
+                // If we have a directory, close all file descriptors except stdin, stdout, stderr and forkPipeOut
+                if (dir)
                 {
-                    foreach (i; 0 .. maxToClose)
+                    scope(exit) closedir(dir);
+
+                    int opendirfd = dirfd(dir);
+
+                    // Iterate over all file descriptors
+                    while(true)
                     {
-                        // don't close pipe write end
-                        if (pfds[i].fd == forkPipeOut) continue;
-                        // POLLNVAL will be set if the file descriptor is invalid.
-                        if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                        dirent* entry = readdir(dir);
+
+                        if (entry is null) break;
+                        if (entry.d_name[0] == '.') continue;
+
+                        int fd = atoi(cast(char*)entry.d_name);
+
+                        // Don't close stdin, stdout, stderr, forkPipeOut or the directory file descriptor
+                        if (fd < 3 || fd == forkPipeOut || fd == opendirfd) continue;
+
+                        close(fd);
                     }
                 }
                 else
                 {
-                    // Fall back to closing everything.
-                    foreach (i; 3 .. maxDescriptors)
+                    bool fallback = true;
+
+                    // This is going to allocate 8 bytes for each possible file descriptor from 3 to r.rlim_cur
+                    if (r.rlim_cur <= 128*1024)
+                    {
+                        // NOTE: malloc() and getrlimit() are not on the POSIX async
+                        // signal safe functions list, but practically this should
+                        // not be a problem. Java VM and CPython also use malloc()
+                        // in its own implementation via opendir().
+                        import core.stdc.stdlib : malloc;
+                        import core.sys.posix.poll : pollfd, poll, POLLNVAL;
+
+                        immutable maxDescriptors = cast(int) r.rlim_cur;
```

Comment:
Move this outside the if statement, as the fallback needs it too.

Reply:
Whooooooops 👍🏻